packages/beacon-node/src/network/reqresp/handlers/dataColumnSidecarsByRange.ts

```diff
@@ -0,0 +1,138 @@
+        columns
+      );
+    }
+  }
+
+  if (endSlot < finalizedSlot) {
+    return;
+  }
+
+  // Non-finalized range of columns
+  const headRoot = chain.forkChoice.getHeadRoot();
+  // TODO DENEB: forkChoice should maintain an array of canonical blocks, and change only on reorg
+  const headChain = chain.forkChoice.getAllAncestorBlocks(headRoot);
+
+  // TODO(fulu): This may cause a lot of overhead in long periods of non-finality.  Need to tune this further
+  // Iterate head chain with ascending block numbers
+  for (let i = headChain.length - 1; i >= 0; i--) {
+    const block = headChain[i];
+
+    // Must include only blobs in the range requested
+    if (block.slot >= startSlot && block.slot < endSlot) {
+      // Note: Here the forkChoice head may change due to a re-org, so the headChain reflects the canonical chain
+      // at the time of the start of the request. Spec is clear the chain of blobs must be consistent, but on
+      // re-org there's no need to abort the request
+      // Spec: https://github.com/ethereum/consensus-specs/blob/a1e46d1ae47dd9d097725801575b46907c12a1f8/specs/eip4844/p2p-interface.md#blobssidecarsbyrange-v1
+
+      const blobSideCarsBytesWrapped = await unfinalized.getBinary(fromHex(block.blockRoot));
+      if (!blobSideCarsBytesWrapped) {
+        // console.log(`error onDataColumnSidecarsByRange No item for root ${block.blockRoot} slot ${block.slot}`);
+        // Handle the same to onBeaconBlocksByRange
+        throw new ResponseError(RespStatus.SERVER_ERROR, `No item for root ${block.blockRoot} slot ${block.slot}`);
+      }
+      yield* iterateDataColumnBytesFromWrapper(chain, blobSideCarsBytesWrapped, block.slot, columns);
+    }
+
+    // If block is after endSlot, stop iterating
+    else if (block.slot >= endSlot) {
+      break;
+    }
+  }
+}
+
+export function* iterateDataColumnBytesFromWrapper(
+  chain: IBeaconChain,
+  dataColumnSidecarsBytesWrapped: Uint8Array,
+  blockSlot: Slot,
+  columns: ColumnIndex[]
+): Iterable<ResponseOutgoing> {
+  const {columnSizeInBytes, custodyIndex, serializedColumnSidecars} =
+    parseWrappedColumnSidecars(dataColumnSidecarsBytesWrapped);
+
+  // no columns possibly no blob
+  if (serializedColumnSidecars.length === 0) {
+    return;
+  }
+
+  const fork = chain.config.getForkName(blockSlot);
+
+  for (const columnIndex of columns) {
+    const dataIndex = getIndexOfSidecarInWrapper(custodyIndex, columnIndex);
+    const dataColumnSidecarBytes = serializedColumnSidecars.slice(
+      dataIndex * columnSizeInBytes,
+      (dataIndex + 1) * columnSizeInBytes
+    );
+    if (dataColumnSidecarBytes.length !== columnSizeInBytes) {
+      throw new ResponseError(
+        RespStatus.SERVER_ERROR,
+        `Invalid dataColumnSidecar columnIndex=${columnIndex} dataIndex=${dataIndex} bytes length=${dataColumnSidecarBytes.length} expected=${columnSizeInBytes} for slot ${blockSlot}`
+      );
+    }
+    yield {
+      data: dataColumnSidecarBytes,
```

Comment:
add util to `sszBytes` to extract `DataColumnSidecar.index` from Uint8Array and confirm it's the same to the requested column index before returning to peers